src/cockpit/router.py

```diff
@@ -26,12 +27,47 @@
 logger = logging.getLogger(__name__)
 
 
+class ExecutionQueue(collections.deque):
+    def run(self):
+        for method, args in self:
+            method(*args)
+
+    def wrap(self, method):
+        """Temporarily delay calls to a given class method.
+
+        Functions by replacing the named function at the instance __dict__
+        level, effectively providing an override for exactly one instance.
+        Queues the invocations.  Run them later with .run(). Reverse this call
+        by deleting the named methods from the instance.
+        """
+        # NB: this function is stored in the instance dict and therefore
+        # doesn't function as a descriptor, isn't a method, doesn't get bound,
+        # and therefore doesn't receive a self parameter
+        setattr(method.__self__, method.__func__.__name__, lambda *args: self.append((method, args)))
+
+
 class Endpoint:
     router: Router
+    queue: Optional[ExecutionQueue] = None
 
     def __init__(self, router: Router):
         self.router = router
 
+    def freeze_endpoint(self):
+        self.queue = ExecutionQueue()
+        self.queue.wrap(self.do_channel_control)
+        self.queue.wrap(self.do_channel_data)
+
+    def thaw_endpoint(self):
+        queue = self.queue
+
+        if queue is not None:
+            self.queue = None
+            queue.run()
+            # remove the enqueue wrappers — will be handled according to class default now
+            del self.do_channel_control
+            del self.do_channel_data
```

Comment:
This is super ugly and error prone -- couldn't this be done by `ExecutionQueue.__del__` or `.run()`? If you can't figure out `self` from `self.method`, then we could additionally pass `self` to the `.wrap()` call and record it, so that run() can remove the wrappers by itself.

Reply:
Ya.  I'm not sure I agree with this.  The queue object doesn't contain any direct references to the frozen endpoint object or the methods (other than via the queued calls, of which there may be none).  We could add that, but it would be added complexity.  Also: doing things as side-effect of `__del__` is definitely evil.

You might convince me to transform the queue into a context manager which takes an object and a list of methods and which we explicitly __exit__...

Reply:
It seems `method` is sufficient -- `method.__self__` is the object of method, I tried that in a REPL.
```
>>> class Foo:
...   def __init__(self, val):
...     self.val = val
...   def say(self):
...     print("hi", hash(self), self.val)
... 

>>> a = Foo(1)
>>> b = Foo(2)

>>> a.say()
hi 8738851316177 1
>>> b.say()
hi 8738851200769 2

>>> a
<__main__.Foo object at 0x7f2ac20e1d10>
>>> b
<__main__.Foo object at 0x7f2ac1f1f010>

>>> a.say.__self__
<__main__.Foo object at 0x7f2ac20e1d10>
>>> b.say.__self__
<__main__.Foo object at 0x7f2ac1f1f010>
```

Reply:
I thought about changing def wrap() to do something like `self.wrapped.append(method)`, and then in run() it can iterate over `self.wrapped` and call del on all of them?

Reply:
This works. I'll let you do the context manager thing and then we can see which we like better.

```diff
--- src/cockpit/router.py
+++ src/cockpit/router.py
@@ -28,9 +28,15 @@ logger = logging.getLogger(__name__)
 
 
 class ExecutionQueue(collections.deque):
+    def __init__(self):
+        super().__init__()
+        self.wrappers = []
+
     def run(self):
         for method, args in self:
             method(*args)
+        for wrapper in self.wrappers:
+            del wrapper.__self__.__dict__[wrapper.__func__.__name__]
 
     def wrap(self, method):
         """Temporarily delay calls to a given class method.
@@ -43,6 +49,7 @@ class ExecutionQueue(collections.deque):
         # NB: this function is stored in the instance dict and therefore
         # doesn't function as a descriptor, isn't a method, doesn't get bound,
         # and therefore doesn't receive a self parameter
+        self.wrappers.append(method)
         setattr(method.__self__, method.__func__.__name__, lambda *args: self.append((method, args)))
 
 
@@ -64,9 +71,6 @@ class Endpoint:
         if queue is not None:
             self.queue = None
             queue.run()
-            # remove the enqueue wrappers — will be handled according to class default now
-            del self.do_channel_control
-            del self.do_channel_data
 
     # interface for receiving messages
     def do_channel_control(self, channel: str, command: str, message: Dict[str, object]) -> None:
```